File: binary_tree_level_order_traversal.js
// Given a binary tree, return the level order traversal of its nodes' values.
//   (ie, from left to right, level by level).
// 
// For example:
// Given binary tree [3,9,20,null,null,15,7],
//     3
//    / \
//   9  20
//     /  \
//    15   7
// return its level order traversal as:
// [
//   [3],
//   [9,20],
//   [15,7]
// ]

var levelOrder = function(root) {
  if (root === null) { return []; }
  var result = [];
  var queue = [root];
  var row;
  var node;
  var left;
  var right;
  var size;

  while (queue.length) {
    row = [];
    size = queue.length;
    
    for (var i = 0; i < size; i++) {
      node = queue.shift();
      row.push(node.val);
      left = node.left;
      right = node.right;

      if (left) { queue.push(left); }
      if (right) { queue.push(right); }
    }

    result.push(row);
  }

  return result;
};

File: binary_tree_level_order_traversal.rb
# Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).
# 
# For example:
# Given binary tree [3,9,20,null,null,15,7],
#     3
#    / \
#   9  20
#     /  \
#    15   7
# return its level order traversal as:
# [
#   [3],
#   [9,20],
#   [15,7]
# ]

def level_order(root)
  return [] if root.nil?

  result = []
  queue = [root]
  current_level = 0

  until queue.empty?
    queue.size.times do
      node = queue.shift
      result[current_level] = [] unless result[current_level]
      result[current_level] << node.val
      queue.push(node.left) if node.left
      queue.push(node.right) if node.right
    end
    current_level += 1
  end

  result
end

